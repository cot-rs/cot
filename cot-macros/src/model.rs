use cot_codegen::model::{Field, Model, ModelArgs, ModelOpts, ModelType};
use darling::FromMeta;
use darling::ast::NestedMeta;
use heck::ToSnakeCase;
use proc_macro2::{Ident, TokenStream};
use quote::{ToTokens, TokenStreamExt, format_ident, quote};
use syn::Token;
use syn::punctuated::Punctuated;

use crate::cot_ident;

#[must_use]
pub(super) fn impl_model_for_struct(
    args: &[NestedMeta],
    ast: &mut syn::DeriveInput,
) -> TokenStream {
    let args = match ModelArgs::from_list(args) {
        Ok(v) => v,
        Err(e) => {
            return e.write_errors();
        }
    };

    let opts = match ModelOpts::new_from_derive_input(ast) {
        Ok(val) => val,
        Err(err) => {
            return err.write_errors();
        }
    };

    let model = match opts.as_model(&args) {
        Ok(val) => val,
        Err(err) => {
            return err.to_compile_error();
        }
    };
    let builder = ModelBuilder::from_model(model);

    let attrs = &ast.attrs;
    let vis = &ast.vis;
    let ident = &ast.ident;

    // Filter out our helper attributes so they don't get passed to the struct
    let fields = match &mut ast.data {
        syn::Data::Struct(data) => &mut data.fields,
        _ => panic!("Only structs are supported"),
    };
    let fields = remove_helper_field_attributes(fields);

    quote!(
        #(#attrs)*
        #vis struct #ident {
            #fields
        }
        #builder
    )
}

fn remove_helper_field_attributes(fields: &mut syn::Fields) -> &Punctuated<syn::Field, Token![,]> {
    match fields {
        syn::Fields::Named(fields) => {
            for field in &mut fields.named {
                field.attrs.retain(|a| !a.path().is_ident("model"));
            }
            &fields.named
        }
        _ => panic!("Only named fields are supported"),
    }
}

#[derive(Debug)]
struct ModelBuilder {
    app_name: String,
    name: Ident,
    vis: syn::Visibility,
    table_name: String,
    pk_field: Field,
    fields_struct_name: Ident,
    fields_as_columns: Vec<TokenStream>,
    fields_as_from_db: Vec<TokenStream>,
    fields_as_update_from_db: Vec<TokenStream>,
    fields_as_get_values: Vec<TokenStream>,
    fields_as_field_refs: Vec<TokenStream>,
}

impl ToTokens for ModelBuilder {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        tokens.append_all(self.build_model_impl());
        tokens.append_all(self.build_fields_struct());
    }
}

impl ModelBuilder {
    fn from_model(model: Model) -> Self {
        let field_count = model.field_count();
        let app_name = std::env::var("CARGO_PKG_NAME").unwrap();
        let table_name = match model.model_type {
            ModelType::Internal => model.table_name,
            _ => format!("{}__{}", app_name.to_snake_case(), model.table_name.to_snake_case()),
        };
        let mut model_builder = Self {
            app_name: app_name,
            name: model.name.clone(),
            vis: model.vis,
            table_name: table_name,
            pk_field: model.pk_field.clone(),
            fields_struct_name: format_ident!("{}Fields", model.name),
            fields_as_columns: Vec::with_capacity(field_count),
            fields_as_from_db: Vec::with_capacity(field_count),
            fields_as_update_from_db: Vec::with_capacity(field_count),
            fields_as_get_values: Vec::with_capacity(field_count),
            fields_as_field_refs: Vec::with_capacity(field_count),
        };
        for field in &model.fields {
            model_builder.push_field(field);
        }

        model_builder
    }

    fn push_field(&mut self, field: &Field) {
        let orm_ident = orm_ident();

        let name = &field.field_name;
        let ty = &field.ty;
        let index = self.fields_as_columns.len();
        let column_name = &field.column_name;

        {
            let field_as_column = quote!(#orm_ident::Column::new(
                #orm_ident::Identifier::new(#column_name)
            ));
            self.fields_as_columns.push(field_as_column);
        }

        self.fields_as_from_db.push(quote!(
            #name: db_row.get::<#ty>(#index)?
        ));

        self.fields_as_update_from_db.push(quote!(
            #index => { self.#name = db_row.get::<#ty>(row_field_id)?; }
        ));

        self.fields_as_get_values.push(quote!(
            #index => &self.#name as &dyn #orm_ident::ToDbFieldValue
        ));

        self.fields_as_field_refs.push(quote!(
            #[doc = concat!("Field reference to [`", stringify!(#name), "::", stringify!(#column_name), "`].")]
            pub const #name: #orm_ident::query::FieldRef<#ty> =
                #orm_ident::query::FieldRef::<#ty>::new(#orm_ident::Identifier::new(#column_name));
        ));
    }

    #[must_use]
    fn build_model_impl(&self) -> TokenStream {
        let crate_ident = cot_ident();
        let orm_ident = orm_ident();

        let name = &self.name;
        let app_name = &self.app_name;
        let table_name = &self.table_name;
        let fields_struct_name = &self.fields_struct_name;
        let fields_as_columns = &self.fields_as_columns;
        let pk_field_name = &self.pk_field.field_name;
        let pk_column_name = &self.pk_field.column_name;
        let pk_type = &self.pk_field.ty;
        let fields_as_from_db = &self.fields_as_from_db;
        let fields_as_update_from_db = &self.fields_as_update_from_db;
        let fields_as_get_values = &self.fields_as_get_values;

        quote! {
            #[#crate_ident::__private::async_trait]
            #[automatically_derived]
            impl #orm_ident::Model for #name {
                type Fields = #fields_struct_name;
                type PrimaryKey = #pk_type;

                const COLUMNS: &'static [#orm_ident::Column] = &[
                    #(#fields_as_columns,)*
                ];
                const APP_NAME: &'static str = #app_name;
                const TABLE_NAME: #orm_ident::Identifier = #orm_ident::Identifier::new(#table_name);
                const PRIMARY_KEY_NAME: #orm_ident::Identifier = #orm_ident::Identifier::new(#pk_column_name);

                fn primary_key(&self) -> &Self::PrimaryKey {
                    &self.#pk_field_name
                }

                fn set_primary_key(&mut self, primary_key: Self::PrimaryKey) {
                    self.#pk_field_name = primary_key;
                }

                fn from_db(db_row: #orm_ident::Row) -> #orm_ident::Result<Self> {
                    Ok(Self {
                        #(#fields_as_from_db,)*
                    })
                }

                fn update_from_db(&mut self, db_row: #orm_ident::Row, columns: &[usize]) -> #orm_ident::Result<()> {
                    for (row_field_id, column_id) in columns.into_iter().enumerate() {
                        match *column_id {
                            #(#fields_as_update_from_db,)*
                            _ => panic!("Unknown column index: {}", column_id),
                        }
                    }

                    Ok(())
                }

                fn get_values(&self, columns: &[usize]) -> Vec<&dyn #orm_ident::ToDbFieldValue> {
                    columns
                        .iter()
                        .map(|&column| match column {
                            #(#fields_as_get_values,)*
                            _ => panic!("Unknown column index: {}", column),
                        })
                        .collect()
                }

                async fn get_by_primary_key<DB: #orm_ident::DatabaseBackend>(
                    db: &DB,
                    pk: Self::PrimaryKey,
                ) -> #orm_ident::Result<Option<Self>> {
                    #orm_ident::query!(Self, $#pk_field_name == pk)
                        .get(db)
                        .await
                }
            }
        }
    }

    #[must_use]
    fn build_fields_struct(&self) -> TokenStream {
        let name = &self.name;
        let vis = &self.vis;
        let fields_struct_name = &self.fields_struct_name;
        let fields_as_field_refs = &self.fields_as_field_refs;

        quote! {
            #[doc = concat!("Fields of the model [`", stringify!(#name), "`].")]
            #[derive(::core::fmt::Debug)]
            #vis struct #fields_struct_name;

            #[allow(non_upper_case_globals)]
            impl #fields_struct_name {
                #(#fields_as_field_refs)*
            }
        }
    }
}

#[must_use]
fn orm_ident() -> TokenStream {
    let crate_ident = cot_ident();
    quote! { #crate_ident::db }
}
