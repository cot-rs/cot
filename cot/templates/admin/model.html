{% extends "base.html" %}

{% block title %}{{ model.name() }}{% endblock %}

{% block content -%}
{%- let request = request -%}
{%- let model = model -%}
<div class="model-header">
    <h2>{{ model.name() }}</h2>
    <div class="action-box">
        <a class="btn primary" href="{{ cot::reverse!(request, "create_model_instance", model_name = model.url_name())? }}">Create {{ model.name() }} {% include "icons/plus.svg" %}</a>
    </div>
</div>

<div class="models-wrapper">
    <table class="models">
        <thead>
            <tr>
                <th>Object</th>
                <th>Actions</th>
            </tr>
        </thead>
        <tbody>
        {%- for object in objects -%}
            <tr>
                {%- let edit_link = cot::reverse!(request, "edit_model_instance", model_name = model.url_name(), pk = object.id())? -%}
                {%- let remove_link = cot::reverse!(request, "remove_model_instance", model_name = model.url_name(), pk = object.id())? -%}
                <td><a href="{{ edit_link }}">{{ object.display() }}</a></td>
                <td class="model-actions-cell"><a href="{{ edit_link }}" class="edit-model" title="Edit this {{ model.name() }}">{% include "icons/pencil.svg" %}</a> <a href="{{ remove_link }}" class="remove-model" title="Remove this {{ model.name() }}">{% include "icons/trash.svg" %}</a></td>
            </tr>
        {%- endfor -%}
        </tbody>
    </table>

    <footer>
        Displaying {{ objects.len() }} out of {{ total_object_counts }} {{ model.name() }}{{ total_object_counts|pluralize }}.

        <div class="pagination">
            <select id="page-size-selector">
                {% for option in [10, 20, 30, 40] %}
                    <option value="{{ option }}" {% if option == page_size %}selected{% endif %}>{{ option }}</option>
                {% endfor %}
            </select>

            {% if page > 1 %}
                <a href="?page={{ page - 1 }}&page_size={{ page_size }}" class="btn secondary">Previous</a>
            {% else %}
                <button class="btn disabled">Previous</button>
            {% endif %}

             <span>Page {{page}} of {{total_pages}}</span>

            {% if page < total_pages %}
                <a href="?page={{ page + 1 }}&page_size={{ page_size }}" class="btn secondary">Next</a>
            {% else %}
                <button class="btn disabled">Next</button>
            {% endif %}
        </div>
    </footer>
</div>

<script>
    document.getElementById("page-size-selector").addEventListener("change", function() {
        const pageSize = this.value;
        const url = new URL(window.location.href);
        url.searchParams.set("page_size", pageSize);
        // Reset to page 1 when changing page size
        url.searchParams.set("page", 1);
        window.location.href = url.toString();
    });
</script>

{%- endblock %}
