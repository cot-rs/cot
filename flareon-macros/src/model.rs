use darling::ast::NestedMeta;
use darling::{FromDeriveInput, FromMeta};
use flareon_codegen::model::{Field, Model, ModelArgs, ModelOpts};
use proc_macro2::{Ident, TokenStream};
use quote::{format_ident, quote, ToTokens, TokenStreamExt};

use crate::flareon_ident;

#[must_use]
pub(super) fn impl_model_for_struct(args: &[NestedMeta], ast: &syn::DeriveInput) -> TokenStream {
    let args = match ModelArgs::from_list(args) {
        Ok(v) => v,
        Err(e) => {
            return e.write_errors();
        }
    };

    let opts = match ModelOpts::from_derive_input(ast) {
        Ok(val) => val,
        Err(err) => {
            return err.write_errors();
        }
    };

    let model = match opts.as_model(&args) {
        Ok(val) => val,
        Err(err) => {
            return err.to_compile_error();
        }
    };
    let builder = ModelBuilder::from_model(model);

    quote!(#ast #builder)
}

#[derive(Debug)]
struct ModelBuilder {
    name: Ident,
    table_name: String,
    fields_struct_name: Ident,
    fields_as_columns: Vec<TokenStream>,
    fields_as_from_db: Vec<TokenStream>,
    fields_as_get_values: Vec<TokenStream>,
    fields_as_field_refs: Vec<TokenStream>,
}

impl ToTokens for ModelBuilder {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        tokens.append_all(self.build_model_impl());
        tokens.append_all(self.build_fields_struct());
    }
}

impl ModelBuilder {
    fn from_model(model: Model) -> Self {
        let field_count = model.field_count();
        let mut model_builder = Self {
            name: model.name.clone(),
            table_name: model.table_name,
            fields_struct_name: format_ident!("{}Fields", model.name),
            fields_as_columns: Vec::with_capacity(field_count),
            fields_as_from_db: Vec::with_capacity(field_count),
            fields_as_get_values: Vec::with_capacity(field_count),
            fields_as_field_refs: Vec::with_capacity(field_count),
        };
        for field in &model.fields {
            model_builder.push_field(field);
        }

        model_builder
    }

    fn push_field(&mut self, field: &Field) {
        let orm_ident = orm_ident();

        let name = &field.field_name;
        let ty = &field.ty;
        let index = self.fields_as_columns.len();
        let column_name = &field.column_name;
        let is_auto = field.auto_value;
        let is_null = field.null;

        {
            let mut field_as_column = quote!(#orm_ident::Column::new(
                #orm_ident::Identifier::new(#column_name)
            ));
            if is_auto {
                field_as_column.append_all(quote!(.auto()));
            }
            if is_null {
                field_as_column.append_all(quote!(.null()));
            }
            self.fields_as_columns.push(field_as_column);
        }

        self.fields_as_from_db.push(quote!(
            #name: db_row.get::<#ty>(#index)?
        ));

        self.fields_as_get_values.push(quote!(
            #index => &self.#name as &dyn #orm_ident::ToDbValue
        ));

        self.fields_as_field_refs.push(quote!(
            pub const #name: #orm_ident::query::FieldRef<#ty> =
                #orm_ident::query::FieldRef::<#ty>::new(#orm_ident::Identifier::new(#column_name));
        ));
    }

    #[must_use]
    fn build_model_impl(&self) -> TokenStream {
        let orm_ident = orm_ident();

        let name = &self.name;
        let table_name = &self.table_name;
        let fields_struct_name = &self.fields_struct_name;
        let fields_as_columns = &self.fields_as_columns;
        let fields_as_from_db = &self.fields_as_from_db;
        let fields_as_get_values = &self.fields_as_get_values;

        quote! {
            #[automatically_derived]
            impl #orm_ident::Model for #name {
                type Fields = #fields_struct_name;

                const COLUMNS: &'static [#orm_ident::Column] = &[
                    #(#fields_as_columns,)*
                ];
                const TABLE_NAME: #orm_ident::Identifier = #orm_ident::Identifier::new(#table_name);

                fn from_db(db_row: #orm_ident::Row) -> #orm_ident::Result<Self> {
                    Ok(Self {
                        #(#fields_as_from_db,)*
                    })
                }

                fn get_values(&self, columns: &[usize]) -> Vec<&dyn #orm_ident::ToDbValue> {
                    columns
                        .iter()
                        .map(|&column| match column {
                            #(#fields_as_get_values,)*
                            _ => panic!("Unknown column index: {}", column),
                        })
                        .collect()
                }
            }
        }
    }

    #[must_use]
    fn build_fields_struct(&self) -> TokenStream {
        let fields_struct_name = &self.fields_struct_name;
        let fields_as_field_refs = &self.fields_as_field_refs;

        quote! {
            #[derive(::core::fmt::Debug)]
            pub struct #fields_struct_name;

            #[allow(non_upper_case_globals)]
            impl #fields_struct_name {
                #(#fields_as_field_refs)*
            }
        }
    }
}

#[must_use]
fn orm_ident() -> TokenStream {
    let crate_ident = flareon_ident();
    quote! { #crate_ident::db }
}
