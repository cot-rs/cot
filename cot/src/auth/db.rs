//! Database-backed user authentication backend.
//!
//! This module provides a user type and an authentication backend that stores
//! the user data in a database using the Cot ORM.

use std::any::Any;

use async_trait::async_trait;
use hmac::{Hmac, KeyInit, Mac};
use sha2::Sha512;
use thiserror::Error;

use crate::admin::{AdminModel, AdminModelManager, DefaultAdminModelManager};
use crate::auth::{
    AuthBackend, AuthError, Password, PasswordHash, PasswordVerificationResult, Result,
    SessionAuthHash, User, UserId,
};
use crate::config::SecretKey;
use crate::db::migrations::SyncDynMigration;
use crate::db::{model, query, Auto, DatabaseBackend, LimitedString, Model};
use crate::request::{Request, RequestExt};
use crate::CotApp;

pub mod migrations;

pub(crate) const MAX_USERNAME_LENGTH: u32 = 255;

/// A user stored in the database.
#[derive(Debug, Clone)]
#[model]
pub struct DatabaseUser {
    id: Auto<i64>,
    #[model(unique)]
    username: LimitedString<MAX_USERNAME_LENGTH>,
    password: PasswordHash,
}

#[derive(Debug, Clone, Error)]
#[non_exhaustive]
pub enum CreateUserError {
    #[error("username is too long (max {MAX_USERNAME_LENGTH} characters, got {0})")]
    UsernameTooLong(usize),
}

impl DatabaseUser {
    #[must_use]
    fn new(
        id: Auto<i64>,
        username: LimitedString<MAX_USERNAME_LENGTH>,
        password: &Password,
    ) -> Self {
        Self {
            id,
            username,
            password: PasswordHash::from_password(password),
        }
    }

    /// Create a new user and save it to the database.
    ///
    /// # Errors
    ///
    /// Returns an error if the user could not be saved to the database.
    ///
    /// # Example
    ///
    /// ```
    /// use cot::auth::db::DatabaseUser;
    /// use cot::auth::Password;
    /// use cot::request::{Request, RequestExt};
    /// use cot::response::{Response, ResponseExt};
    /// use cot::{Body, StatusCode};
    ///
    /// async fn view(request: &Request) -> cot::Result<Response> {
    ///     let user = DatabaseUser::create_user(
    ///         request.db(),
    ///         "testuser".to_string(),
    ///         &Password::new("password123"),
    ///     )
    ///     .await?;
    ///
    ///     Ok(Response::new_html(
    ///         StatusCode::OK,
    ///         Body::fixed("User created!"),
    ///     ))
    /// }
    ///
    /// # #[tokio::main]
    /// # async fn main() -> cot::Result<()> {
    /// #     use cot::test::{TestDatabase, TestRequestBuilder};
    /// #     let mut test_database = TestDatabase::new_sqlite().await?;
    /// #     test_database.with_auth().run_migrations().await;
    /// #     let request = TestRequestBuilder::get("/")
    /// #         .with_db_auth(test_database.database())
    /// #         .build();
    /// #     view(&request).await?;
    /// #     test_database.cleanup().await?;
    /// #     Ok(())
    /// # }
    /// ```
    pub async fn create_user<DB: DatabaseBackend, T: Into<String>, U: Into<Password>>(
        db: &DB,
        username: T,
        password: U,
    ) -> Result<Self> {
        let username = username.into();
        let username_length = username.len();
        let username = LimitedString::<MAX_USERNAME_LENGTH>::new(username).map_err(|_| {
            AuthError::backend_error(CreateUserError::UsernameTooLong(username_length))
        })?;

        let mut user = Self::new(Auto::auto(), username, &password.into());
        user.save(db).await.map_err(AuthError::backend_error)?;

        Ok(user)
    }

    pub async fn get_by_id<DB: DatabaseBackend>(db: &DB, id: UserId) -> Result<Option<Self>> {
        let id = id.as_int().expect("User ID should be an integer");

        let db_user = query!(DatabaseUser, $id == id)
            .get(db)
            .await
            .map_err(AuthError::backend_error)?;

        Ok(db_user)
    }

    pub async fn get_by_username<DB: DatabaseBackend>(
        db: &DB,
        username: &str,
    ) -> Result<Option<Self>> {
        let username = LimitedString::<MAX_USERNAME_LENGTH>::new(username).map_err(|_| {
            AuthError::backend_error(CreateUserError::UsernameTooLong(username.len()))
        })?;
        let db_user = query!(DatabaseUser, $username == username)
            .get(db)
            .await
            .map_err(AuthError::backend_error)?;

        Ok(db_user)
    }

    pub async fn authenticate<DB: DatabaseBackend>(
        db: &DB,
        credentials: &DatabaseUserCredentials,
    ) -> Result<Option<Self>> {
        let username = credentials.username();
        let username_limited = LimitedString::<MAX_USERNAME_LENGTH>::new(username.to_string())
            .map_err(|_| {
                AuthError::backend_error(CreateUserError::UsernameTooLong(username.len()))
            })?;
        let user = query!(DatabaseUser, $username == username_limited)
            .get(db)
            .await
            .map_err(AuthError::backend_error)?;

        if let Some(mut user) = user {
            let password_hash = &user.password;
            match password_hash.verify(credentials.password()) {
                PasswordVerificationResult::Ok => Ok(Some(user)),
                PasswordVerificationResult::OkObsolete(new_hash) => {
                    user.password = new_hash;
                    user.save(db).await.map_err(AuthError::backend_error)?;
                    Ok(Some(user))
                }
                PasswordVerificationResult::Invalid => Ok(None),
            }
        } else {
            // SECURITY: If no user was found, run the same hashing function to prevent
            // timing attacks from being used to determine if a user exists. Additionally,
            // do something with the result to prevent the compiler from optimizing out the
            // operation.
            // TODO: benchmark this to make sure it works as expected
            let dummy_hash = PasswordHash::from_password(credentials.password());
            if let PasswordVerificationResult::Invalid = dummy_hash.verify(credentials.password()) {
                panic!("Password hash verification should never fail for a newly generated hash");
            }
            Ok(None)
        }
    }

    #[must_use]
    pub fn id(&self) -> i64 {
        self.id
            .expect("called DatabaseUser::id() on an unsaved instance")
    }

    #[must_use]
    pub fn username(&self) -> &str {
        &self.username
    }
}

type SessionAuthHmac = Hmac<Sha512>;

impl User for DatabaseUser {
    fn id(&self) -> Option<UserId> {
        Some(UserId::Int(self.id()))
    }

    fn username(&self) -> Option<&str> {
        Some(&self.username)
    }

    fn is_active(&self) -> bool {
        true
    }

    fn is_authenticated(&self) -> bool {
        true
    }

    fn session_auth_hash(&self, secret_key: &SecretKey) -> Option<SessionAuthHash> {
        let mut mac = SessionAuthHmac::new_from_slice(secret_key.as_bytes())
            .expect("HMAC can take key of any size");
        mac.update(self.password.as_str().as_bytes());
        let hmac_data = mac.finalize().into_bytes();

        Some(SessionAuthHash::new(&hmac_data))
    }
}

#[async_trait]
impl AdminModel for DatabaseUser {
    async fn get_objects(request: &Request) -> crate::Result<Vec<Self>> {
        Ok(DatabaseUser::objects()
            .all(request.db())
            .await
            .map_err(AuthError::backend_error)?)
    }

    fn name() -> &'static str {
        "DatabaseUser"
    }

    fn url_name() -> &'static str {
        "database_user"
    }

    fn display(&self) -> String {
        format!("{self:?}")
    }
}

/// Credentials for authenticating a user stored in the database.
///
/// This struct is used to authenticate a user stored in the database. It
/// contains the username and password of the user.
///
/// Can be passed to
/// [`AuthRequestExt::authenticate`](crate::auth::AuthRequestExt::authenticate)
/// to authenticate a user when using the [`DatabaseUserBackend`].
#[derive(Debug, Clone)]
pub struct DatabaseUserCredentials {
    username: String,
    password: Password,
}

impl DatabaseUserCredentials {
    #[must_use]
    pub fn new(username: String, password: Password) -> Self {
        Self { username, password }
    }

    #[must_use]
    pub fn username(&self) -> &str {
        &self.username
    }

    #[must_use]
    pub fn password(&self) -> &Password {
        &self.password
    }
}

/// The authentication backend for users stored in the database.
///
/// This is the default authentication backend for Cot. It authenticates
/// users stored in the database using the [`DatabaseUser`] model.
///
/// This backend supports authenticating users using the
/// [`DatabaseUserCredentials`] struct and ignores all other credential types.
#[derive(Debug, Copy, Clone)]
pub struct DatabaseUserBackend;

impl Default for DatabaseUserBackend {
    fn default() -> Self {
        Self::new()
    }
}

impl DatabaseUserBackend {
    /// Create a new instance of the database user authentication backend.
    ///
    /// # Example
    ///
    /// ```
    /// use cot::auth::db::DatabaseUserBackend;
    /// use cot::config::ProjectConfig;
    ///
    /// let backend = DatabaseUserBackend::new();
    /// let config = ProjectConfig::builder().auth_backend(backend).build();
    /// ```
    #[must_use]
    pub fn new() -> Self {
        Self {}
    }
}

#[async_trait]
impl AuthBackend for DatabaseUserBackend {
    async fn authenticate(
        &self,
        request: &Request,
        credentials: &(dyn Any + Send + Sync),
    ) -> Result<Option<Box<dyn User + Send + Sync>>> {
        if let Some(credentials) = credentials.downcast_ref::<DatabaseUserCredentials>() {
            #[allow(trivial_casts)] // Upcast to the correct Box type
            Ok(DatabaseUser::authenticate(request.db(), credentials)
                .await
                .map(|user| user.map(|user| Box::new(user) as Box<dyn User + Send + Sync>))?)
        } else {
            Ok(None)
        }
    }

    async fn get_by_id(
        &self,
        request: &Request,
        id: UserId,
    ) -> Result<Option<Box<dyn User + Send + Sync>>> {
        #[allow(trivial_casts)] // Upcast to the correct Box type
        Ok(DatabaseUser::get_by_id(request.db(), id)
            .await?
            .map(|user| Box::new(user) as Box<dyn User + Send + Sync>))
    }
}

#[derive(Debug, Copy, Clone)]
pub struct DatabaseUserApp;

impl Default for DatabaseUserApp {
    fn default() -> Self {
        Self::new()
    }
}

impl DatabaseUserApp {
    #[must_use]
    pub fn new() -> Self {
        Self {}
    }
}

impl CotApp for DatabaseUserApp {
    fn name(&self) -> &'static str {
        "cot_db_user"
    }

    fn admin_model_managers(&self) -> Vec<Box<dyn AdminModelManager>> {
        vec![Box::new(DefaultAdminModelManager::<DatabaseUser>::new())]
    }

    fn migrations(&self) -> Vec<Box<SyncDynMigration>> {
        // TODO: this is way too complicated for the user-facing API
        #[allow(trivial_casts)]
        migrations::MIGRATIONS
            .iter()
            .copied()
            .map(|x| Box::new(x) as Box<SyncDynMigration>)
            .collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::SecretKey;
    use crate::db::MockDatabaseBackend;

    #[test]
    #[cfg_attr(miri, ignore)]
    fn session_auth_hash() {
        let user = DatabaseUser::new(
            Auto::fixed(1),
            LimitedString::new("testuser").unwrap(),
            &Password::new("password123"),
        );
        let secret_key = SecretKey::new(b"supersecretkey");

        let hash = user.session_auth_hash(&secret_key);
        assert!(hash.is_some());
    }

    #[test]
    #[cfg_attr(miri, ignore)]
    fn database_user_traits() {
        let user = DatabaseUser::new(
            Auto::fixed(1),
            LimitedString::new("testuser").unwrap(),
            &Password::new("password123"),
        );
        let user_ref: &dyn User = &user;
        assert_eq!(user_ref.id(), Some(UserId::Int(1)));
        assert_eq!(user_ref.username(), Some("testuser"));
        assert!(user_ref.is_active());
        assert!(user_ref.is_authenticated());
        assert!(user_ref
            .session_auth_hash(&SecretKey::new(b"supersecretkey"))
            .is_some());
    }

    #[tokio::test]
    #[cfg_attr(miri, ignore)]
    async fn create_user() {
        let mut mock_db = MockDatabaseBackend::new();
        mock_db
            .expect_insert::<DatabaseUser>()
            .returning(|_| Ok(()));

        let username = "testuser".to_string();
        let password = Password::new("password123");

        let user = DatabaseUser::create_user(&mock_db, username.clone(), &password)
            .await
            .unwrap();
        assert_eq!(user.username(), username);
    }

    #[tokio::test]
    #[cfg_attr(miri, ignore)]
    async fn get_by_id() {
        let mut mock_db = MockDatabaseBackend::new();
        let user = DatabaseUser::new(
            Auto::fixed(1),
            LimitedString::new("testuser").unwrap(),
            &Password::new("password123"),
        );

        mock_db
            .expect_get::<DatabaseUser>()
            .returning(move |_| Ok(Some(user.clone())));

        let result = DatabaseUser::get_by_id(&mock_db, UserId::Int(1))
            .await
            .unwrap();
        assert!(result.is_some());
        assert_eq!(result.unwrap().username(), "testuser");
    }

    #[tokio::test]
    #[cfg_attr(miri, ignore)]
    async fn authenticate() {
        let mut mock_db = MockDatabaseBackend::new();
        let user = DatabaseUser::new(
            Auto::fixed(1),
            LimitedString::new("testuser").unwrap(),
            &Password::new("password123"),
        );

        mock_db
            .expect_get::<DatabaseUser>()
            .returning(move |_| Ok(Some(user.clone())));

        let credentials =
            DatabaseUserCredentials::new("testuser".to_string(), Password::new("password123"));
        let result = DatabaseUser::authenticate(&mock_db, &credentials)
            .await
            .unwrap();
        assert!(result.is_some());
        assert_eq!(result.unwrap().username(), "testuser");
    }

    #[tokio::test]
    #[cfg_attr(miri, ignore)]
    async fn authenticate_non_existing() {
        let mut mock_db = MockDatabaseBackend::new();

        mock_db
            .expect_get::<DatabaseUser>()
            .returning(move |_| Ok(None));

        let credentials =
            DatabaseUserCredentials::new("testuser".to_string(), Password::new("password123"));
        let result = DatabaseUser::authenticate(&mock_db, &credentials)
            .await
            .unwrap();
        assert!(result.is_none());
    }

    #[tokio::test]
    #[cfg_attr(miri, ignore)]
    async fn authenticate_invalid_password() {
        let mut mock_db = MockDatabaseBackend::new();
        let user = DatabaseUser::new(
            Auto::fixed(1),
            LimitedString::new("testuser").unwrap(),
            &Password::new("password123"),
        );

        mock_db
            .expect_get::<DatabaseUser>()
            .returning(move |_| Ok(Some(user.clone())));

        let credentials =
            DatabaseUserCredentials::new("testuser".to_string(), Password::new("invalid"));
        let result = DatabaseUser::authenticate(&mock_db, &credentials)
            .await
            .unwrap();
        assert!(result.is_none());
    }
}
