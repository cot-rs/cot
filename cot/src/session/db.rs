//! Database-backed session management backend.
//!
//! This module provides a session type and an app for storing session data
//! in a database using the Cot ORM.
pub mod migrations;

use cot::db::migrations::SyncDynMigration;

use crate::App;
use crate::db::{Auto, model};

/// Session data stored in the database.
#[derive(Debug, Clone)]
#[model]
pub struct Session {
    #[model(primary_key)]
    pub(crate) id: Auto<i32>,
    #[model(unique)]
    pub(crate) key: String,
    pub(crate) data: String,
    pub(crate) expiry: chrono::DateTime<chrono::FixedOffset>,
}

/// An app that provides session management via a session model stored in the
/// database.
///
/// This app registers the session model and its migrations, enabling persistent
/// session storage in the database.
///
/// # Examples
///
/// ```no_run
/// use cot::config::{DatabaseConfig, ProjectConfig};
/// use cot::project::RegisterAppsContext;
/// use cot::session::db::SessionApp;
/// use cot::{App, AppBuilder, Project};
///
/// struct MyProject;
/// impl Project for MyProject {
///     fn config(&self, config_name: &str) -> cot::Result<ProjectConfig> {
///         Ok(ProjectConfig::builder()
///             .database(DatabaseConfig::builder().url("sqlite::memory:").build())
///             .build())
///     }
///
///     fn register_apps(&self, apps: &mut AppBuilder, _context: &RegisterAppsContext) {
///         apps.register_with_views(SessionApp::new(), "");
///     }
/// }
///
/// #[cot::main]
/// fn main() -> impl Project {
///     MyProject
/// }
/// ```

#[derive(Debug, Copy, Clone)]
#[non_exhaustive]
pub struct SessionApp;

impl SessionApp {
    /// Create a new instance of the session management app.
    ///
    /// # Examples
    ///
    /// ```
    /// use cot::session::db::SessionApp;
    /// let app = SessionApp::new();
    /// ```
    #[must_use]
    pub fn new() -> Self {
        Self {}
    }
}

impl Default for SessionApp {
    fn default() -> Self {
        Self::new()
    }
}

impl App for SessionApp {
    fn name(&self) -> &'static str {
        "cot_session"
    }

    fn migrations(&self) -> Vec<Box<SyncDynMigration>> {
        cot::db::migrations::wrap_migrations(migrations::MIGRATIONS)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::App;

    #[test]
    fn test_session_app_basic_behavior() {
        let app1 = SessionApp::new();

        let app2 = SessionApp::default();

        assert_eq!(app1.name(), "cot_session");
        assert_eq!(app2.name(), "cot_session");

        let migrations = app1.migrations();
        assert!(!migrations.is_empty());
    }
}
