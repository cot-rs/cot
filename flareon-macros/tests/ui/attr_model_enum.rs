use flareon::db::model;

#[model]
enum MyModel {
    A(i32),
    B(std::string::String),
    C(String),
    D(i32),
}

fn main() {}
