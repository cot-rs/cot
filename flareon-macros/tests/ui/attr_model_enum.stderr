error: Unsupported shape `enum`. Expected struct with named fields.
 --> tests/ui/attr_model_enum.rs:3:1
  |
3 | #[model]
  | ^^^^^^^^
  |
  = note: this error originates in the attribute macro `model` (in Nightly builds, run with -Z macro-backtrace for more info)
